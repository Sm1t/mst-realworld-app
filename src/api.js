import axios from 'axios';

axios.defaults.baseURL = 'https://conduit.productionready.io/api';

export const ArticleService = {
  all: () => axios.get('/articles'),
  get: slug => axios.get(`/articles/${slug}`),
};