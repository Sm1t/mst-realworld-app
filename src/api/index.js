import axios from 'axios';

axios.defaults.baseURL = 'https://conduit.productionready.io/api';
// eslint-disable-next-line dot-notation
// axios.defaults.headers.common['Authorization'] = 'AUTH_TOKEN';

export * from './article.service';
export * from './auth.service';
export * from './tag.service';
