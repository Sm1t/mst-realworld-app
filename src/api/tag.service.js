import axios from 'axios';

export const TagService = {
  getAll: () => axios.get('/tags').then(({ data: { tags } }) => tags),
};
