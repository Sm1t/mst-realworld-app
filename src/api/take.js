const take = what => res => res.data[what];

export default take;