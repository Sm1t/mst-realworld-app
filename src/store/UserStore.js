import { types, flow } from 'mobx-state-tree';
import Router from 'next/router';

import { AuthService } from '../api';

const User = types.model('User', {
  bio: types.string,
  createdAt: types.string,
  email: types.string,
  id: types.number,
  image: types.string,
  token: types.string,
  updatedAt: types.string,
  username: types.string,
});

export const UserStore = types
  .model('UserStore', {
    currentUser: types.maybeNull(User),
    errors: types.array(types.string),
  })
  .actions(self => ({
    setCurrentUser: user => {
      self.currentUser = user;
    },
    login: flow(function* login(email, password) {
      self.errors.clear();
      try {
        const user = yield AuthService.login(email, password);

        self.currentUser = user;
        Router.replace('/');
      } catch (error) {
        const errors =
          error.response && error.response.data && error.response.data.errors;

        if (errors) {
          self.errors = Object.keys(errors).map(key => `${key} ${errors[key]}`);
        }
      }
    }),
    checkAuth: flow(function* checkAuth(token) {
      try {
        const user = yield AuthService.current(token);

        self.currentUser = user;
      } catch (error) {} // eslint-disable-line no-empty
    }),
  }));
