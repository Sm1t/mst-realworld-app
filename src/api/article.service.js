import axios from 'axios';

import { ARTICLES } from './constants';
import take from './take';

export const ArticleService = {
  all: () => axios.get(`/${ARTICLES}`).then(take('articles')),
  get: slug => axios.get(`/${ARTICLES}/${slug}`).then(take('article')),
};
