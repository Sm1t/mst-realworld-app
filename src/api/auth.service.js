import axios from 'axios';
import cookies from 'js-cookie';

import take from './take';

export const AuthService = {
  current: token => axios
    .get('/user', token ? { headers: { Authorization: token } } : undefined)
    .then(take('user')),

  login(email, password) {
    return axios
      .post('/users/login', { user: { email, password } })
      .then(take('user'))
      .then(user => {
        this.setToken(user.token);

        return user;
      });
  },

  getToken: () => cookies.get('token'),

  setToken: t => {
    const token = `Token ${t}`;

    cookies.set('token', token);
    // eslint-disable-next-line dot-notation
    axios.defaults.headers.common['Authorization'] = token;
  },

  removeToken: () => {
    cookies.remove('token');
    // eslint-disable-next-line dot-notation
    delete axios.defaults.headers.common['Authorization'];
  },
};
