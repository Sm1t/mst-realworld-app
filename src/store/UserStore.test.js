import { AuthService } from '../api';

import { UserStore } from './UserStore';

jest.mock('../api/auth.service');

describe('UserStore', () => {
  const user = {
    id: 1,
    email: 'test@test.ru',
    createdAt: '2019-03-23T22:28:55.669Z',
    updatedAt: '2019-04-21T05:01:38.095Z',
    username: 'test',
    bio: 'Lorem ipsum dolor sip amet',
    image: 'https://path.to/img',
    token: 'token',
  };

  it('should create store without problems', () => {
    const userStore = UserStore.create();

    expect(userStore).toMatchSnapshot();
  });

  describe('actions', () => {
    let userStore;

    beforeEach(() => {
      userStore = UserStore.create();
      AuthService.login.mockReset();
    });

    it('should properly set currentUser', () => {
      userStore.setCurrentUser(user);
      expect(userStore).toMatchSnapshot();
    });

    it('should login', async () => {
      AuthService.login.mockResolvedValue(user);
      await userStore.login(user.email, 'password');

      expect(AuthService.login).toHaveBeenCalledWith(user.email, 'password');
      expect(userStore.currentUser).toEqual(user);
    });

    it('should handle login errors', async () => {
      AuthService.login.mockRejectedValue({
        response: {
          data: {
            errors: {
              'email or password': 'is invalid',
            },
          },
        },
      });
      await userStore.login();
      expect(userStore.errors).toEqual(['email or password is invalid']);
    });
  });
});
