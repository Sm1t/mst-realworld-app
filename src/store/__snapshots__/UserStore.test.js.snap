// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserStore actions should properly set currentUser 1`] = `
Object {
  "currentUser": Object {
    "bio": "Lorem ipsum dolor sip amet",
    "createdAt": "2019-03-23T22:28:55.669Z",
    "email": "test@test.ru",
    "id": 1,
    "image": "https://path.to/img",
    "token": "token",
    "updatedAt": "2019-04-21T05:01:38.095Z",
    "username": "test",
  },
  "errors": Array [],
}
`;

exports[`UserStore should create store without problems 1`] = `
Object {
  "currentUser": null,
  "errors": Array [],
}
`;
